src/overlays/actors/ovl_En_Tp/z_en_tp.c

```diff
@@ -58,43 +79,699 @@ static ColliderJntSphInit D_80B22AB4 = {
+ */
+void EnTp_Die(EnTp* this, GlobalContext* globalCtx) {
+    EnTp* now;
+    s16 i;
+    s32 pad;
+    Vec3f effectVelAccel = { 0.0f, 0.5f, 0.0f };
+    Vec3f effectPos = { 0.0f, 0.0f, 0.0f };
+
+    this->timer--;
+
+    if (this->timer <= 0) {
+        if (this->actor.params == TAILPASARAN_HEAD_DYING) {
+            effectPos.x = ((Rand_ZeroOne() - 0.5f) * 15.0f) + this->actor.world.pos.x;
+            effectPos.z = ((Rand_ZeroOne() - 0.5f) * 15.0f) + this->actor.world.pos.z;
+            effectPos.y = ((Rand_ZeroOne() - 0.5f) * 5.0f) + this->actor.world.pos.y;
+            EffectSsDeadDb_Spawn(globalCtx, &effectPos, &effectVelAccel, &effectVelAccel, 100, 0, 255, 255, 255, 255, 0,
+                                 0, 255, 1, 9, 1);
+
+            effectPos.x = ((Rand_ZeroOne() - 0.5f) * 15.0f) + this->actor.world.pos.x;
+            effectPos.z = ((Rand_ZeroOne() - 0.5f) * 15.0f) + this->actor.world.pos.z;
+            effectPos.y = ((Rand_ZeroOne() - 0.5f) * 5.0f) + this->actor.world.pos.y;
+            EffectSsDeadDb_Spawn(globalCtx, &effectPos, &effectVelAccel, &effectVelAccel, 100, 0, 255, 255, 255, 255, 0,
+                                 0, 255, 1, 9, 1);
+            Item_DropCollectibleRandom(globalCtx, &this->actor, &this->actor.world.pos, 0x50);
+        } else {
+            for (i = 0; i < 1; i++) {
+                now =
+                    (EnTp*)Actor_Spawn(&globalCtx->actorCtx, globalCtx, ACTOR_EN_TP, this->actor.world.pos.x,
+                                       this->actor.world.pos.y, this->actor.world.pos.z, 0, 0, 0, TAILPASARAN_FRAGMENT);
+
+                if (now != NULL) {
+                    Actor_SetScale(&now->actor, this->actor.scale.z * 0.5f);
+                    now->red = this->red;
+                }
+            }
+        }
+
+        if (this->actor.child != NULL) {
+            this->actor.child->parent = NULL;
+            this->actor.child->params = TAILPASARAN_TAIL_DYING;
+            this->actor.child->colChkInfo.health = 0;
+        }
+
+        this->unk_150 = 2;
+        Actor_Kill(&this->actor);
+    }
+}
+
+void EnTp_Fragment_SetupFade(EnTp* this) {
+    this->actionIndex = TAILPASARAN_ACTION_FRAGMENT_FADE;
+    this->actor.world.pos.x += ((Rand_ZeroOne() - 0.5f) * 5.0f);
+    this->actor.world.pos.y += ((Rand_ZeroOne() - 0.5f) * 5.0f);
+    this->actor.world.pos.z += ((Rand_ZeroOne() - 0.5f) * 5.0f);
+    this->actor.velocity.x = (Rand_ZeroOne() - 0.5f) * 1.5f;
+    this->actor.velocity.y = (Rand_ZeroOne() - 0.5f) * 1.5f;
+    this->actor.velocity.z = (Rand_ZeroOne() - 0.5f) * 1.5f;
+    this->actor.flags &= ~1;
+    EnTp_SetupAction(this, EnTp_Fragment_Fade);
+}
+
+void EnTp_Fragment_Fade(EnTp* this, GlobalContext* globalCtx) {
+    func_8002D7EC(&this->actor);
+    this->alpha -= 20;
+
+    if (this->alpha < 20) {
+        this->alpha = 0;
+        Actor_Kill(&this->actor);
+    }
+}
+
+void EnTp_Head_SetupTakeOff(EnTp* this) {
+    this->timer = (s32)((Rand_ZeroOne() * 15.0f) + 40.0f);
```

Comment:
Is this cast necessary?

Reply:
I thought it might be, since timer is an s16, but apparently not.